File: app/src/main/java/net/twisterrob/android/view/color/ColorPickerView.java
package net.twisterrob.android.view.color;

import java.util.*;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.*;
import android.os.Build;
import android.util.*;
import android.view.*;
import android.widget.ImageView;

import static android.util.TypedValue.*;

import net.twisterrob.android.view.color.swatches.*;
import net.twisterrob.android.view.color.swatches.pixel.color.*;
import net.twisterrob.android.view.color.swatches.pixel.drawer.*;

public class ColorPickerView extends ImageView implements SwatchChooser.OnSwatchChangeListener {
	public interface OnColorChangedListener {
		void colorChanged(int color);
	}

	private OnColorChangedListener mListener;
	private final Touchy touch;
	private Swatch swatch;

	public ColorPickerView(Context context) {
		super(context);
		touch = new Touchy(context);
		init(context);
	}

	public ColorPickerView(Context context, AttributeSet attrs) {
		super(context, attrs);
		touch = new Touchy(context);
		init(context);
	}

	public ColorPickerView(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		touch = new Touchy(context);
		init(context);
	}

	private void init(Context context) {
		setOnLongClickListener(touch);
		setOnTouchListener(touch);
		setFocusableInTouchMode(true);
		setupDefaultSwatch();
	}

	private void setupDefaultSwatch() {
		Swatch swatch = swatches.iterator().next();
		if (!isInEditMode()) {
			if (swatch instanceof PixelAbsoluteSwatch) {
				((PixelAbsoluteSwatch)swatch).forceAsync();
			}
		}
		setSwatch(swatch);
	}

	public OnColorChangedListener getColorChangedListener() {
		return mListener;
	}

	public void setColorChangedListener(OnColorChangedListener listener) {
		this.mListener = listener;
	}

	public boolean isContinousMode() {
		return touch.isContinuousMode();
	}

	public void setContinousMode(boolean continousMode) {
		touch.setContinuousMode(continousMode);
	}

	public Swatch getSwatch() {
		return swatch;
	}

	public boolean setSwatch(int swatchIndex) {
		if (0 <= swatchIndex && swatchIndex < swatches.size()) {
			setSwatch(swatches.get(swatchIndex));
			return true;
		}
		return false;
	}

	public void setSwatch(Swatch swatch) {
		if (swatch == null) {
			throw new NullPointerException("Swatch cannot be null");
		}
		if (this.swatch != null) {
			swatch.setCurrentColor(this.swatch.getCurrentColor());
		}
		this.swatch = swatch;
		setImageDrawable(swatch);
	}

	public int getColor() {
		return swatch.getCurrentColor();
	}

	public void setColor(int color) {
		if (color == swatch.getCurrentColor()) {
			return;
		}
		swatch.setCurrentColor(color);
		invalidate();
		fireColorChanged();
	}

	/**
	 * @see <a href="http://www.jayway.com/2012/12/12/creating-custom-android-views-part-4-measuring-and-how-to-force-a-view-to-be-square">Tutorial</a>
	 */
	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		super.onMeasure(widthMeasureSpec, heightMeasureSpec);
		final float RATIO = 1f / 1f;

		int width = getMeasuredWidth();
		int height = getMeasuredHeight();
		int widthWithoutPadding = width - getPaddingLeft() - getPaddingRight();
		int heightWithoutPadding = height - getPaddingTop() - getPaddingBottom();

		int maxWidth = (int)(heightWithoutPadding * RATIO);
		int maxHeight = (int)(widthWithoutPadding / RATIO);

		if (maxWidth < maxHeight) {
			width = maxWidth + getPaddingLeft() + getPaddingRight();
		} else {
			height = maxHeight + getPaddingTop() + getPaddingBottom();
		}

		setMeasuredDimension(width, height);
	}

	@TargetApi(Build.VERSION_CODES.HONEYCOMB)
	@Override
	public int getBaseline() {
		if (Build.VERSION_CODES.HONEYCOMB <= Build.VERSION.SDK_INT && getBaselineAlignBottom()) {
			return getMeasuredHeight();
		}
		return getPaddingTop() + (getMeasuredHeight() - getPaddingTop() - getPaddingBottom()) / 2;
	}

	protected void fireColorChanged() {
		if (mListener != null) {
			mListener.colorChanged(swatch.getCurrentColor());
		}
	}

	private final class Touchy implements OnTouchListener, OnLongClickListener {
		private final int mTouchSlop;
		private int trackedArea = Swatch.AREA_INVALID;
		private int highlightArea = Swatch.AREA_INVALID;
		private boolean continuousMode = true;

		public Touchy(Context context) {
			mTouchSlop = ViewConfiguration.get(context).getScaledTouchSlop();
		}

		public boolean isContinuousMode() {
			return continuousMode;
		}

		public void setContinuousMode(boolean continuousMode) {
			this.continuousMode = continuousMode;
		}

		private PointF longTapStart = new PointF();

		@Override
		public boolean onTouch(View v, MotionEvent event) {
			if (getDrawable() instanceof SwatchChooser) {
				return ((SwatchChooser)getDrawable()).onTouch(v, event);
			}
			invalidate();
			float x = event.getX() - getPaddingLeft();
			float y = event.getY() - getPaddingTop();
			if (!swatch.getBounds().contains((int)x, (int)y)) {
				return unhandled(event);
			}
			int inCenter = swatch.getAreaCode(x, y);

			switch (event.getAction()) {
				case MotionEvent.ACTION_DOWN:
					startLongTap(x, y);
					//rebuild();
					trackedArea = inCenter;
					if (trackedArea != Swatch.AREA_INVALID && !continuousMode) {
						swatch.setCurrentArea(highlightArea = inCenter);
					}
					return handled(event);
				case MotionEvent.ACTION_MOVE:
					updateLongTap(x, y);
					if (trackedArea != Swatch.AREA_INVALID) {
						if (trackedArea == inCenter && !continuousMode) {
							swatch.setCurrentArea(highlightArea = inCenter);
						} else {
							swatch.setCurrentArea(highlightArea = Swatch.AREA_INVALID);
						}
						int color = swatch.findColor(trackedArea, x, y);
						if (continuousMode) {
							setColor(color);
						} else {
							swatch.setCurrentColor(color);
						}
					}
					return handled(event);
				case MotionEvent.ACTION_UP:
					if (trackedArea != Swatch.AREA_INVALID) {
						if (!continuousMode && trackedArea == highlightArea
								&& swatch.triggersColorChange(trackedArea)) {
							int color = swatch.findColor(trackedArea, x, y);
							setColor(color);
							View view = ColorPickerView.this; // ClickableViewAccessibility doesn't see the connection
							view.performClick();
						}
						trackedArea = Swatch.AREA_INVALID;
						swatch.setCurrentArea(highlightArea = Swatch.AREA_INVALID);
					}
					return handled(event);
				case MotionEvent.ACTION_CANCEL:
					trackedArea = Swatch.AREA_INVALID;
					swatch.setCurrentArea(highlightArea = Swatch.AREA_INVALID);
					return handled(event);
			}
			return unhandled(event);
		}

		private boolean unhandled(MotionEvent event) {
			return false;
		}

		public boolean handled(MotionEvent event) {
			onTouchEvent(event);
			return true;
		}

		public void startLongTap(float x, float y) {
			longTapStart.set(x, y);
		}

		private void updateLongTap(float x, float y) {
			float dist2 = (x - longTapStart.x) * (x - longTapStart.x) + (y - longTapStart.y) * (y - longTapStart.y);
			if (Math.sqrt(dist2) >= mTouchSlop) {
				cancelLongPress();
			}
		}

		@Override
		public boolean onLongClick(View v) {
			showChooser();
			return true;
		}
	}

	@Override
	public boolean performClick() {
		if (swatch != null) {
			if (swatch instanceof PixelAbsoluteSwatch) {
				((PixelAbsoluteSwatch)swatch).forceAsync();
			}
			setSwatch(swatch);
		}
		return super.performClick();
	}

	private final List<Swatch> swatches = new ArrayList<>(Arrays.asList(
			new PixelAbsoluteSwatch(ColorReplacer.wrap(CenterBitmapDrawer.factory(), Color.BLACK, Color.TRANSPARENT),
					new RadialHSBGradient()),
			new PixelAbsoluteSwatch(ColumnByColumnBitmapDrawer.factory(), new LinearHSBGradient()),
			new APIDemoSwatch(),
			new PixelAbsoluteSwatch(CenterBitmapDrawer.factory(), new RadialHueGradient()),
			new PixelAbsoluteSwatch(LineByLineBitmapDrawer.factory(), new LinearHueGradient())
	));

	/**
	 * Manipulate as you wish.
	 */
	public List<Swatch> getSwatches() {
		return swatches;
	}

	/**
	 * To hide chooser: call {@link #setSwatch}.
	 */
	public void showChooser() {
		SwatchChooser chooser = new SwatchChooser(swatches);
		chooser.setOnSwatchChangeListener(this);
		chooser.setTileMargin((int)TypedValue.applyDimension(COMPLEX_UNIT_DIP, 4, getResources().getDisplayMetrics()));
		setImageDrawable(chooser);
	}

	@Override
	public void swatchSelected(Swatch swatch) {
		this.swatch = swatch;
	}
}

File: feature/base/src/main/java/net/twisterrob/android/view/color/ColorPickerView.kt
package net.twisterrob.android.view.color

import android.annotation.TargetApi
import android.content.Context
import android.graphics.Color
import android.graphics.PointF
import android.os.Build
import android.support.v7.widget.AppCompatImageView
import android.util.AttributeSet
import android.util.TypedValue
import android.util.TypedValue.COMPLEX_UNIT_DIP
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import net.twisterrob.android.view.color.swatches.APIDemoSwatch
import net.twisterrob.android.view.color.swatches.PixelAbsoluteSwatch
import net.twisterrob.android.view.color.swatches.Swatch
import net.twisterrob.android.view.color.swatches.pixel.color.ColorReplacer
import net.twisterrob.android.view.color.swatches.pixel.color.LinearHSBGradient
import net.twisterrob.android.view.color.swatches.pixel.color.LinearHueGradient
import net.twisterrob.android.view.color.swatches.pixel.color.RadialHSBGradient
import net.twisterrob.android.view.color.swatches.pixel.color.RadialHueGradient
import net.twisterrob.android.view.color.swatches.pixel.drawer.CenterBitmapDrawer
import net.twisterrob.android.view.color.swatches.pixel.drawer.ColumnByColumnBitmapDrawer
import net.twisterrob.android.view.color.swatches.pixel.drawer.LineByLineBitmapDrawer
import kotlin.math.sqrt

open class ColorPickerView : AppCompatImageView, SwatchChooser.OnSwatchChangeListener {

	interface OnColorChangedListener {
		fun colorChanged(color: Int)
	}

	var colorChangedListener: OnColorChangedListener? = null
	private val touch: Touchy
	private lateinit var swatch: Swatch

	constructor(context: Context) : super(context) {
		touch = Touchy(context)
		init(context)
	}

	constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
		touch = Touchy(context)
		init(context)
	}

	constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
		touch = Touchy(context)
		init(context)
	}

	private fun init(@Suppress("UNUSED_PARAMETER") context: Context) {
		setOnLongClickListener(touch)
		setOnTouchListener(touch)
		isFocusableInTouchMode = true
		setupDefaultSwatch()
	}

	private fun setupDefaultSwatch() {
		val swatch = swatches.iterator().next()
		if (!isInEditMode) {
			(swatch as? PixelAbsoluteSwatch)?.forceAsync()
		}
		setSwatch(swatch)
	}

	var isContinuousMode: Boolean // === by touch::isContinuousMode
		get() = touch.isContinuousMode
		set(continuousMode) {
			touch.isContinuousMode = continuousMode
		}

	fun getSwatch(): Swatch? {
		return swatch
	}

	fun setSwatch(swatchIndex: Int): Boolean {
		if (0 <= swatchIndex && swatchIndex < swatches.size) {
			setSwatch(swatches[swatchIndex])
			return true
		}
		return false
	}

	fun setSwatch(swatch: Swatch) {
		if (this::swatch.isInitialized) {
			swatch.currentColor = this.swatch.currentColor
		}
		this.swatch = swatch
		setImageDrawable(swatch)
	}

	var color: Int // by swatch.currentColor
		get() = swatch.currentColor
		set(color) {
			if (color == swatch.currentColor) {
				return
			}
			swatch.currentColor = color
			invalidate()
			fireColorChanged()
		}

	/**
	 * @see [Tutorial](http://www.jayway.com/2012/12/12/creating-custom-android-views-part-4-measuring-and-how-to-force-a-view-to-be-square)
	 */
	override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
		super.onMeasure(widthMeasureSpec, heightMeasureSpec)
		val RATIO = 1f / 1f

		var width = measuredWidth
		var height = measuredHeight
		val widthWithoutPadding = width - paddingLeft - paddingRight
		val heightWithoutPadding = height - paddingTop - paddingBottom

		val maxWidth = (heightWithoutPadding * RATIO).toInt()
		val maxHeight = (widthWithoutPadding / RATIO).toInt()

		if (maxWidth < maxHeight) {
			width = maxWidth + paddingLeft + paddingRight
		} else {
			height = maxHeight + paddingTop + paddingBottom
		}

		setMeasuredDimension(width, height)
	}

	@TargetApi(Build.VERSION_CODES.HONEYCOMB)
	override fun getBaseline(): Int {
		if (Build.VERSION_CODES.HONEYCOMB <= Build.VERSION.SDK_INT && baselineAlignBottom) {
			return measuredHeight
		}
		return paddingTop + (measuredHeight - paddingTop - paddingBottom) / 2
	}

	protected open fun fireColorChanged() {
		colorChangedListener?.colorChanged(swatch.currentColor)
	}

	private inner class Touchy(context: Context) : View.OnTouchListener, View.OnLongClickListener {

		private val mTouchSlop: Int = ViewConfiguration.get(context).scaledTouchSlop
		private var trackedArea = Swatch.AREA_INVALID
		private var highlightArea = Swatch.AREA_INVALID
			set(value) {
				field = value
				swatch.setCurrentArea(field)
			}
		var isContinuousMode = true

		private val longTapStart = PointF()

		override fun onTouch(v: View, event: MotionEvent): Boolean {
			if (drawable is SwatchChooser) {
				return (drawable as SwatchChooser).onTouch(v, event)
			}
			invalidate()
			val x = event.x - paddingLeft
			val y = event.y - paddingTop
			if (!swatch.bounds.contains(x.toInt(), y.toInt())) {
				return unhandled(event)
			}
			val inCenter = swatch.getAreaCode(x, y)

			when (event.action) {
				MotionEvent.ACTION_DOWN -> {
					startLongTap(x, y)
					//rebuild()
					trackedArea = inCenter
					if (trackedArea != Swatch.AREA_INVALID && !isContinuousMode) {
						highlightArea = inCenter
					}
					return handled(event)
				}

				MotionEvent.ACTION_MOVE -> {
					updateLongTap(x, y)
					if (trackedArea != Swatch.AREA_INVALID) {
						if (trackedArea == inCenter && !isContinuousMode) {
							highlightArea = inCenter
						} else {
							highlightArea = Swatch.AREA_INVALID
						}
						val color = swatch.findColor(trackedArea, x, y)
						if (isContinuousMode) {
							this@ColorPickerView.color = color
						} else {
							swatch.currentColor = color
						}
					}
					return handled(event)
				}

				MotionEvent.ACTION_UP -> {
					if (trackedArea != Swatch.AREA_INVALID) {
						if (!isContinuousMode
							&& trackedArea == highlightArea
							&& swatch.triggersColorChange(trackedArea)
						) {
							val color = swatch.findColor(trackedArea, x, y)
							this@ColorPickerView.color = color
							val view = this@ColorPickerView // ClickableViewAccessibility doesn't see the connection
							view.performClick()
						}
						trackedArea = Swatch.AREA_INVALID
						highlightArea = Swatch.AREA_INVALID
					}
					return handled(event)
				}

				MotionEvent.ACTION_CANCEL -> {
					trackedArea = Swatch.AREA_INVALID
					highlightArea = Swatch.AREA_INVALID
					return handled(event)
				}
			}
			return unhandled(event)
		}

		private fun unhandled(@Suppress("UNUSED_PARAMETER") event: MotionEvent) = false

		private fun handled(event: MotionEvent) = true.also {
			onTouchEvent(event)
		}

		fun startLongTap(x: Float, y: Float) {
			longTapStart.set(x, y)
		}

		private fun updateLongTap(x: Float, y: Float) {
			val dist2 = (x - longTapStart.x) * (x - longTapStart.x) + (y - longTapStart.y) * (y - longTapStart.y)
			if (sqrt(dist2) >= mTouchSlop) {
				cancelLongPress()
			}
		}

		override fun onLongClick(v: View) = true.also {
			showChooser()
		}
	}

	override fun performClick(): Boolean {
		(swatch as? PixelAbsoluteSwatch)?.forceAsync()
		setSwatch(swatch)
		return super.performClick()
	}

	/**
	 * Manipulate as you wish.
	 */
	val swatches: List<Swatch> = mutableListOf(
		PixelAbsoluteSwatch(
			ColorReplacer.wrap(CenterBitmapDrawer.factory(), Color.BLACK, Color.TRANSPARENT),
			RadialHSBGradient()
		),
		PixelAbsoluteSwatch(ColumnByColumnBitmapDrawer.factory(), LinearHSBGradient()),
		APIDemoSwatch(),
		PixelAbsoluteSwatch(CenterBitmapDrawer.factory(), RadialHueGradient()),
		PixelAbsoluteSwatch(LineByLineBitmapDrawer.factory(), LinearHueGradient())
	)

	val swatchIndex get() = this.swatches.indexOf(this.swatch)

	/**
	 * To hide chooser: call [swatch::set].
	 */
	fun showChooser() {
		val chooser = SwatchChooser(swatches)
		chooser.setOnSwatchChangeListener(this)
		chooser.setTileMargin(TypedValue.applyDimension(COMPLEX_UNIT_DIP, 4f, resources.displayMetrics).toInt())
		setImageDrawable(chooser)
	}

	override fun swatchSelected(swatch: Swatch) {
		this.swatch = swatch
	}
}